pyro/poutine/collapse_messenger.py

```diff
@@ -109,8 +109,13 @@ def _process_message(self, msg):
     def _pyro_sample(self, msg):
         if self._block:
             return
+        dim_to_name = {f.dim: f.name for f in msg["cond_indep_stack"]}
```

Comment:
@eb8680 I'm not sure we'll have the complete `cond_indep_stack` here. Won't it be missing any plates enclosing the `poutine.collapse`? I'm not sure why tests are passing. Do you think we could handle this by promoting `collapse.__enter__()` to be effectful and recording those plates in a `IndepMessenger._pyro_collapse()` method?

Reply:
We won't have access to all the plates via `msg["cond_indep_stack"]` at this point, but I believe all the missing ones are already contained in `self.preserved_plates`?

Reply:
Also, what's the reason we need to compute `to_funsor(msg["fn"])` in `_pyro_sample` rather than `_pyro_post_sample`?  Is it just so that we can compute `domain`? If so, would it be easier to leave `msg["fn"]` to `_pyro_post_sample` and compute `domain` directly in `_pyro_sample`?

Reply:
Good point, I've refactored `self.preserved_plates` to contain needed information, so this should now work.

Correct, the reason we compute `to_funsor(msg["fn"])` in `_pyro_sample()` is indeed to infer value domain. This actually seems like the cheapest surefire way to compute value domain, and we'll need to compute `to_funsor(msg["fn"])` anyway, so I'm happy with that computation living in `_pyro_sample()`. We could move it to `_pyro_post_sample()`, but that seems more complex and equivalent, since in `_pyro_sample()` we set `msg["stop"] = True` and immediately pass control to `_pyro_post_sample()`.